Huobi futures api

lamtrhieu opened the pull request:
Hi @kimquy  @quantd2 
This is just work in progress but I would appreciate if you can spend some times review it first. 
I will try to add more tests tomorrow. 

Thanks

kilu01 commented:
I've merged this branch to my branch so anh @kimquy you just need to review https://github.com/acuityinnovations/huobi.ex/pull/2 which is the common structure that anh @lamtrhieu and me are agree upon

sirius248 commented on lib/ex_huobi/auth.ex:
```diff
@@ -0,0 +1,143 @@
+defmodule Huobi.Test do
```
`Huobi.Test` should be `Huobi.Auth`

lamtrhieu replied:
@kimquy This file should be removed. I use it for testing. I will delete it now.

sirius248 commented on lib/ex_huobi/auth.ex:
```diff
@@ -0,0 +1,143 @@
+      DateTime.utc_now() |> IO.inspect
+      |> DateTime.truncate(:second) |> IO.inspect
+      |> DateTime.to_iso8601() |> IO.inspect
+      |> String.split_at(-1) |> IO.inspect
```
Pls remove `IO.inspect`

sirius248 commented on lib/ex_huobi/auth.ex:
```diff
@@ -0,0 +1,143 @@
+  # def sign(method, host, path, params) do
+
+  def create() do
+    method = "POST\n"
```
pls remove `\n`

sirius248 commented on lib/ex_huobi/auth.ex:
```diff
@@ -0,0 +1,143 @@
+    endpoint =
+      ("https://api.hbdm.com/api/v1/contract_order" <>
+         "?" <> params_string <> "&Signature=#{signature}")
+      |> IO.inspect()
```
remove IO.inspect

sirius248 commented on lib/ex_huobi/auth.ex:
```diff
@@ -0,0 +1,143 @@
+
+    signature =
+      :sha256
+      |> :crypto.hmac("94472b76-30a703ac-e75d7396-c34a0", presigned_text)
```
Not sure what is this `"94472b76-30a703ac-e75d7396-c34a0"`

sirius248 commented on lib/ex_huobi/rest/http_client.ex:
```diff
@@ -0,0 +1,78 @@
+  def signed_path(method, host, path, params) do
+    %ExHuobi.Config{api_key: api_key, api_secret: api_secret} = ExHuobi.Config.get(nil)
+
+    api_key = "3c8dd81a-d8b5e97e-qv2d5ctgbn-9bd14"
```
this should not be here

Commit:
add type spec

Commit:
Add function specs for public futures api

Commit:
Add 2 unit tests for create order using VCR

Commit:
Add more unit tests

Commit:
Merge margin-ft into current branch

Commit:
Revert "Merge margin-ft into current branch"

This reverts commit 2ab00a2e2d7bcc0472463443cf025870f657a237, reversing
changes made to 16a99fdf329558e46a572c4d837160318aadac5a.

Commit:
Merge remote-tracking branch 'origin/master' into huobi-futures-api

Commit:
Fix format, credo, dyalyzer

Commit:
Format code again

Commit:
Fix problem when dyalyzer not allow default param config

sirius248 commented on lib/ex_huobi/futures/rest/account.ex:
```diff
@@ -1,17 +1,20 @@
 
-  @type params :: map
-  @type config :: ExHuobi.Config.t()
+  @type config :: ExHuobi.Config.config_or_nil
```
@lamtrhieu this is a bit unconventional.  Would suggest as following

```elixir
@type config :: ExHuobi.Config.t() | nil
```

sirius248 commented on lib/ex_huobi/futures/util.ex:
```diff
@@ -0,0 +1,27 @@
+defmodule ExHuobi.Futures.Util do
```
I know naming is difficult. But in this case, `Util` not the best. can we rename to `ExHuobi.Futures.Handler`

sirius248 commented on lib/ex_huobi/margin/rest/account.ex:
```diff
@@ -4,15 +4,15 @@ defmodule ExHuobi.Margin.Rest.Account do
 
   @type params :: map
-  @type config :: Config.t()
+  @type config :: ExHuobi.Config.config_or_nil
```
Similar here `@type config :: ExHuobi.Config.t() | nil`

sirius248 commented on lib/ex_huobi/margin/rest/order.ex:
```diff
@@ -5,7 +5,7 @@ defmodule ExHuobi.Margin.Rest.Order do
   @type order_id :: integer
   @type params :: map | [map]
-  @type config :: Config.t()
+  @type config :: ExHuobi.Config.config_or_nil
```
`@type config :: ExHuobi.Config.t() | nil`

Commit:
- Rename util class to handler class
- Reuse the same http_client for both futures and margin
- Add document for futures exchange api

Commit:
Fix unformat file

Commit:
Enforce user to add config param when invoke huobi rest api

Commit:
Fix not allow config as nil from public huobi api

sirius248 commented on README.md:
```diff
@@ -64,3 +64,61 @@ ExHuobi.Margin.Rest.Order.bulk_create(
+
+order_id = 1234567890
+
+ExHuobi.Futures.Rest.Order.create_bulk_orders(order_id, config)
```
@lamtrhieu one last small comment. This is order bulk creation, why passing in only 1 single order id ?

lamtrhieu replied:
@kimquy ah, it's a typo. Thanks for spot this out. Let me fix it.

Commit:
fix Readme

kilu01 reviewed (approved):
LGTM
